library/init/data/nat/bitwise.lean

```diff
@@ -23,30 +23,187 @@ namespace nat
+  | 0     := (nat.div_one _).symm
+  | (k+1) := (congr_arg (/ 2) (shiftr_eq_div_pow k)).trans $
+             by dsimp; rw [nat.div_div_eq_div_mul]; refl
+
+  def zero_shiftr (n) : shiftr 0 n = 0 :=
+  (shiftr_eq_div_pow _ _).trans (nat.zero_div _)
+
+  def test_bit_zero (b n) : test_bit (bit b n) 0 = b := bodd_bit _ _
+
+  def test_bit_succ (m b n) : test_bit (bit b n) (succ m) = test_bit n m :=
+  have bodd (shiftr (shiftr (bit b n) 1) m) = bodd (shiftr n m), by rw shiftr1_bit,
+  by rw [-shiftr_add, add_comm] at this; exact this
+
+  def binary_rec {C : nat → Sort u} (f : ∀ b n, C n → C (bit b n)) (z : C 0) : Π n, C n
+  | n := if n0 : n = 0 then by rw n0; exact z else let n' := shiftr n 1 in
     have n' < n, from (div_lt_iff_lt_mul _ _ dec_trivial).2 $
     by note := nat.mul_lt_mul_of_pos_left (dec_trivial : 1 < 2)
          (lt_of_le_of_ne (zero_le _) (ne.symm n0));
        rwa mul_one at this,
+    by rw [-show bit (bodd n) n' = n, from bit_decomp n]; exact 
     f (bodd n) n' (binary_rec n')
 
   def size : ℕ → ℕ := binary_rec (λ_ _, succ) 0
 
   def bits : ℕ → list bool := binary_rec (λb _ IH, b :: IH) []
 
-  def bit (b : bool) : ℕ → ℕ := cond b bit1 bit0
-
   def bitwise (f : bool → bool → bool) : ℕ → ℕ → ℕ :=
   binary_rec
     (λa m Ia, binary_rec
       (λb n _, bit (f a b) (Ia n))
       (cond (f tt ff) (bit a m) 0))
-    (λb, cond (f ff tt) b 0)
+    (λn, cond (f ff tt) n 0)
 
   def lor   : ℕ → ℕ → ℕ := bitwise bor
   def land  : ℕ → ℕ → ℕ := bitwise band
   def ldiff : ℕ → ℕ → ℕ := bitwise (λ a b, a && bnot b)
   def lxor  : ℕ → ℕ → ℕ := bitwise bxor
 
+  set_option type_context.unfold_lemmas true
+  lemma binary_rec_eq {C : nat → Sort u} {f : ∀ b n, C n → C (bit b n)} {z}
+    (h : f ff 0 z = z) (b n) :
+    binary_rec f z (bit b n) = f b n (binary_rec f z n) :=
+  begin
+    rw [binary_rec.equations._eqn_1],
+    cases (by apply_instance : decidable (bit b n = 0)) with b0 b0; dsimp [dite],
+    { generalize (binary_rec._main._pack._proof_2 (bit b n)) e,
+      rw [bodd_bit, shiftr1_bit], intro e, refl },
+    { generalize (binary_rec._main._pack._proof_1 (bit b n) b0) e,
+      note bf := bodd_bit b n, note n0 := shiftr1_bit b n,
+      rw b0 at bf n0, rw [-show ff = b, from bf, -show 0 = n, from n0], intro e,
+      exact h.symm },
+  end
+  
+  lemma binary_rec_zero {C : nat → Sort u} (f : ∀ b n, C n → C (bit b n)) (z) :
+    binary_rec f z 0 = z := rfl
+
+  lemma bitwise_bit_aux {f : bool → bool → bool} (h : f ff ff = ff) :
+    @binary_rec (λ_, ℕ)
+      (λ b n _, bit (f ff b) (cond (f ff tt) n 0))
+      (cond (f tt ff) (bit ff 0) 0) =
+    λ (n : ℕ), cond (f ff tt) n 0 :=
+  begin
+    apply funext, intro n,
+    apply bit_cases_on n, intros b n, rw [binary_rec_eq],
+    { cases b; try {rw h}; ginduction f ff tt with fft; dsimp [cond]; refl },
+    { rw [h, show cond (f ff tt) 0 0 = 0, by cases f ff tt; refl,
+             show cond (f tt ff) (bit ff 0) 0 = 0, by cases f tt ff; refl]; refl }
+  end
```

Comment:
I think there is a misunderstanding here. They may not do what you want, but they do what they are supposed to do.
The type ascription `(t : T)` makes sure that `t` has a type definitionally equal to `T`.
They do **not** guarantee that `t` will have a type structurally equal to `T`.
So, they are not good for `rw` since it relies on the precise structure of the left hand side of the equation.